### PR TITLE
feat(ai): add Google Gemini LLM provider support

### LINKED ISSUE #3693: NestBot(#908): Reaction Management, RAG Search Failures, and Greeting Handling Bugs
## Describe the bug
This is a sub-issue of #908 (NestBot AI Assistant Implementation).

Several bugs and UX issues were discovered during testing of the NestBot AI Assistant:

1. **Reaction Management Issues**: Bot shows  👀 (eyes) and doesn't show ✅ (tick) reactions for completed tasks, or sometimes only 👀 without ever providing an answer. This creates a confusing user experience where reactions remain stuck and bot status is unclear.

2. **RAG Search Failures**: RAG agent returns "I don't have access to information..." even when relevant chunks exist in the database. Users receive unhelpful responses for valid queries that should be answerable.

---

## To Reproduce

1. **Greeting Routing Bug:**
   - Go to Slack channel
   - Send message: "@NestBot Hello"
   - See error: Bot incorrectly routes to agents instead of giving friendly greeting

2. **Reaction Management Bug:**
   - Go to Slack channel
   - Send any query to @NestBot
   - Wait for response
   - See error: Still 👀 reactions appear (should  show ✅)

## Expected behavior

1. **Greeting Handling**: Simple greetings like "Hello", "Hi", "Thanks" should receive friendly, predefined responses without being routed to agents. Only greetings that contain actual questions should be processed by agents.

2. **RAG Search**: RAG agent should try multiple search strategies with different keywords and rephrased queries when initial search yields no results. It should provide helpful guidance when information isn't found, rather than generic "I don't have access" messages.

3. **Reaction Management**: 
   - Only ✅ (white_check_mark) should remain for successful responses (no 👀 reaction should remain)
   - 🤷 (shrugging) reaction should appear when no answer can be generated
   - 👀 reaction should be properly removed in all code paths

4. **Test Coverage**: Comprehensive test queries should cover all intent categories (project, chapter, contribution, RAG) and edge cases (greetings, low confidence queries, error scenarios).

---

## Are you going to work on fixing this?

- [x] Yes
- [ ] No

---

## Screenshots

```
N/A - These are backend/Slack integration bugs that don't have visual screenshots.
```

### PR DESCRIPTION
**Pull request by rajnisk**
Resolves #3693
Related to #908

## Proposed change

This PR adds Google Gemini as an alternative LLM provider for NestBot AI Assistant, enabling contributors to run NestBot locally using a free Google Gemini API key as an alternative to OpenAI.

This is part of the larger NestBot AI Assistant feature work tracked in #908.

### Changes

- **New**: `GoogleEmbedder` class for Google embeddings using `gemini-embedding-001` model (1536 dimensions)
- **Updated**: `llm_config.py` to support provider selection (OpenAI/Google) with automatic fallback
- **Updated**: `embeddings/factory.py` to return Google embedder when `LLM_PROVIDER=google`
- **Added**: Django settings for Gemini (`GOOGLE_API_KEY`, `GOOGLE_MODEL_NAME`, `LLM_PROVIDER`)
- **Added**: Tests for Google LLM provider configuration
- **Updated**: `pyproject.toml` to include `google-genai` extra for CrewAI
- **Updated**: Custom dictionary to include `genai` and `generativeai`
- **Updated**: `.env.example` with new Gemini environment variables

### Technical Details

- Uses `google.genai` SDK (with fallback to deprecated `google.generativeai` for compatibility)
- Implements REST API fallback if SDK is not available
- Default embedding model: `gemini-embedding-001` (recommended, replaces deprecated models)
- Default LLM model: `gemini-2.0-flash`
- Maintains backward compatibility with OpenAI provider (default)

### Configuration

Set these environment variables to use Google provider:

```bash
DJANGO_LLM_PROVIDER=google
DJANGO_GOOGLE_API_KEY=<your-google-api-key>
DJANGO_GOOGLE_MODEL_NAME=gemini-2.0-flash  # Optional, has default
```

For OpenAI provider (default):
```bash
DJANGO_LLM_PROVIDER=openai  # or omit (defaults to openai)
DJANGO_OPEN_AI_SECRET_KEY=<your-openai-key>
```

### Testing

- ✅ All existing tests pass
- ✅ New tests added for Google provider configuration
- ✅ `make check-test-backend` passes locally
- ✅ Tested with Google Gemini API key locally
- ✅ Verified embedding generation works correctly
- ✅ Verified LLM provider selection works as expected

### Files Changed

- `backend/apps/ai/common/llm_config.py` - Provider selection logic
- `backend/apps/ai/embeddings/google.py` - New Google embedder implementation
- `backend/apps/ai/embeddings/factory.py` - Factory updates for Google provider
- `backend/settings/base.py` - New Django settings
- `backend/tests/apps/ai/common/llm_config_test.py` - Test coverage
- `backend/pyproject.toml` - Dependency updates
- `backend/.env.example` - Environment variable documentation
- `cspell/custom-dict.txt` - Spell checker updates

## Checklist

- [x] **Required:** I followed the [contributing workflow](https://github.com/OWASP/Nest/blob/main/CONTRIBUTING.md#contributing-workflow)
- [x] **Required:** I verified that my code works as intended and resolves the issue as described
- [x] **Required:** I ran `make check-test-backend` locally: all warnings addressed, tests passed
- [x] I used AI for code, documentation, tests, or communication related to this PR

### CODE REVIEW

**Comment by rajnisk** [comment]
@coderabbitai resume

File: backend/apps/ai/embeddings/google.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            # Use Google Generative AI SDK (preferred method)
+            # SDK handles batching automatically
+            results = []
+            for text in texts:
+                result = genai.embed_content(
+                    model=self.model,
+                    content=text,
+                )
+                results.append(result["embedding"])
+            return results
```
[the remark this answers is not in the document]
Comment (rajnisk) [reply backend/apps/ai/embeddings/google.py]: @coderabbitai  re-review

[the remark this answers is not in the document]
Comment (rajnisk) [reply backend/apps/ai/embeddings/google.py]: @coderabbitai Fixed in latest commit. The misleading comment has been replaced with accurate comments stating "Process sequentially" on lines 284 and 298. The code now correctly reflects sequential processing.

File: backend/pyproject.toml [hunk 2]
```diff
@@ -14,7 +14,7 @@ packages = [ { include = "apps" } ]
 [tool.poetry.dependencies]
 algoliasearch = "^4.13.2"
 algoliasearch-django = "^4.0.0"
-crewai = { version = "^1.7.2", python = ">=3.10,<3.14" }
+crewai = { version = "^1.7.2", python = ">=3.10,<3.14", extras = [ "google-genai" ] }
```
Comment (rudransh-shrivastava) [inline backend/pyproject.toml]: This changes the lock file, I think you forgot to push it.

File: backend/apps/ai/embeddings/google.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
+"""Google implementation of embedder."""
+
+from __future__ import annotations
+
+from typing import NoReturn
+
+# Try deprecated google.generativeai first (has configure/embed_content methods)
```
Comment (rudransh-shrivastava) [inline backend/apps/ai/embeddings/google.py]: why do we need to use the deprecated API?
I think this adds unnecessary complexity.

File: backend/apps/ai/embeddings/google.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+# This ensures get_dimensions() returns correct values for downstream vector storage
+MODEL_DIMENSIONS: dict[str, int] = {
+    "gemini-embedding-001": 768,  # Recommended, current model
+    "text-embedding-004": 768,  # Deprecated but same dimensions
+    "embedding-001": 768,  # Alternative model
+}
+
+# Default model and dimensions (used as fallback)
+DEFAULT_MODEL = "gemini-embedding-001"
+DEFAULT_DIMENSIONS = 768
```
Comment (rudransh-shrivastava) [inline backend/apps/ai/embeddings/google.py]: Let's use `1536` - similar to OpenAI/`text-embedding-3-small`.

File: backend/apps/ai/embeddings/google.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+        Note:
+            If an unsupported model is provided, a warning is issued and the default
+            model (gemini-embedding-001) is used to ensure correct vector dimensions
+            for downstream storage.
+
+        """
+        self.api_key = settings.GOOGLE_API_KEY
+        # Initialize base_url for REST API fallback (used in all code paths)
+        self.base_url = "https://generativelanguage.googleapis.com/v1beta"
```
Comment (rudransh-shrivastava) [inline backend/apps/ai/embeddings/google.py]: We can use the [SDK](https://ai.google.dev/gemini-api/docs#python) instead of this. 

For context: We are planning to decouple backend and NestBot at the docker and dependency level. Adding it shouldn't be a problem.

**Review by rudransh-shrivastava — CHANGES_REQUESTED** [review]
Hi, thank you for adding this, I have some suggestions:

Also, as a rule of thumb, comments should not explain what the code does. If a comment just restates what's obvious from reading the code, in most cases, it's not needed.

File: backend/apps/ai/embeddings/google.py (hunk 3 above)
Comment (rajnisk) [reply backend/apps/ai/embeddings/google.py]: Thank you for the feedback! You're  right - using deprecated API adds unnecessary complexity. I've removed all deprecated API support and simplified the code to use only `google.genai.Client` SDK. The implementation now follows the same clean pattern as the OpenAI embedder.

File: backend/apps/ai/embeddings/google.py (hunk 4 above)
Comment (rajnisk) [reply backend/apps/ai/embeddings/google.py]: I've updated the dimensions to 1536 to match OpenAI's `text-embedding-3-small`. The code now uses `gemini-embedding-001` model with `output_dimensionality=1536` parameter in the API calls.

File: backend/apps/ai/embeddings/google.py (hunk 5 above)
Comment (rajnisk) [reply backend/apps/ai/embeddings/google.py]: Thank you for the context! I've removed all REST API fallback code and now use only the `google.genai.Client` SDK. The implementation is now much simpler

File: backend/pyproject.toml (hunk 2 above)
Comment (rajnisk) [reply backend/pyproject.toml]: I've now updated the `poetry.lock` file.

Comment (rajnisk) [reply backend/pyproject.toml]: The poetry.lock file has been updated and committed (commit 78e87220). The lock file now includes the google-genai extra dependency.

File: backend/tests/unit/apps/ai/common/llm_config_test.py [hunk 6]
```diff
@@ -3,15 +3,16 @@
 import os
 from unittest.mock import Mock, patch
 
-import pytest
-
 from apps.ai.common.llm_config import get_llm
 
 
 class TestLLMConfig:
     """Test cases for LLM configuration."""
```
Comment (rajnisk) [inline backend/tests/unit/apps/ai/common/llm_config_test.py:13]: Tests are failing because django-configurations isn't picking up the patched environment variables.

Image

File: backend/apps/ai/embeddings/google.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    def __init__(self, model: str = "gemini-embedding-001") -> None:
+        """Initialize Google embedder.
+
+        Args:
+            model: The Google embedding model to use.
+
+        """
+        self.client = genai.Client(api_key=settings.GOOGLE_API_KEY)
+        self.model = model
+        self._dimensions = 1536
```
[the remark this answers is not in the document]
Comment (rajnisk) [reply backend/apps/ai/embeddings/google.py]: @coderrabbitai Normalization is already implemented.

[the remark this answers is not in the document]
Comment (rajnisk) [reply backend/apps/ai/embeddings/google.py]: @coderabbitai[bot] Normalization is already implemented.

File: backend/apps/ai/common/llm_config.py [hunk 8]
```diff
@@ -2,32 +2,46 @@
...
+            api_key=settings.OPEN_AI_SECRET_KEY,
             temperature=0.1,
         )
-    if provider == "anthropic":
+    if provider == "google":
         return LLM(
-            model=os.getenv("ANTHROPIC_MODEL_NAME", "claude-3-5-sonnet-20241022"),
-            api_key=os.getenv("ANTHROPIC_API_KEY"),
+            model=settings.GOOGLE_MODEL_NAME,
+            base_url="https://generativelanguage.googleapis.com/v1beta/openai/",
```
Comment (rudransh-shrivastava) [inline backend/apps/ai/common/llm_config.py]: do we need `base_url` ?
https://docs.crewai.com/en/concepts/llms#google-gemini-api

Example code from docs:
```python
from crewai import LLM

llm = LLM(
    model="gemini/gemini-2.0-flash",
    api_key="your-api-key",  # Or set GOOGLE_API_KEY/GEMINI_API_KEY
    temperature=0.7
)
```

File: backend/settings/base.py [hunk 9]
```diff
@@ -218,7 +218,18 @@ class Base(Configuration):
 
     STATIC_ROOT = BASE_DIR / "staticfiles"
 
-    OPEN_AI_SECRET_KEY = values.SecretValue(environ_name="OPEN_AI_SECRET_KEY")
+    # django-configurations automatically prefixes with "DJANGO_" and uppercases,
+    # so OPEN_AI_SECRET_KEY becomes DJANGO_OPEN_AI_SECRET_KEY (which is what all
+    # tests and code references use). No need to specify environ_name explicitly.
```
Comment (rudransh-shrivastava) [inline backend/settings/base.py]: these comments are unnecessary.

File: backend/settings/base.py [hunk 10]
```diff
@@ -218,7 +218,18 @@ class Base(Configuration):
...
+    OPEN_AI_SECRET_KEY = values.SecretValue()
+    OPEN_AI_MODEL_NAME = values.Value(default="gpt-4o-mini")
```
Comment (rudransh-shrivastava) [inline backend/settings/base.py]: adding a model name seems confusing because it's unclear where this particular model is used. Let's just stick to the existing approach.

File: backend/settings/base.py [hunk 11]
```diff
@@ -218,7 +218,18 @@ class Base(Configuration):
...
+    # Note: GOOGLE_API_KEY uses Value() instead of SecretValue() because it's optional
+    # (only required when LLM_PROVIDER == "google"). SecretValue() requires the env var
+    # to always be set, which breaks setups using only OpenAI. This should still be
+    # treated as a secret and not exposed in logs or configuration output.
+    GOOGLE_API_KEY = values.Value(default=None)
+    GOOGLE_MODEL_NAME = values.Value(default="gemini-2.0-flash")
```
Comment (rudransh-shrivastava) [inline backend/settings/base.py]: same here.

File: backend/apps/ai/common/llm_config.py [hunk 12]
```diff
@@ -2,32 +2,46 @@
...
 def get_llm() -> LLM:
     """Get configured LLM instance.
 
     Returns:
-        LLM: Configured LLM instance with gpt-4.1-mini as default model.
+        LLM: Configured LLM instance based on settings.
 
     """
-    provider = os.getenv("LLM_PROVIDER", "openai")
+    provider = settings.LLM_PROVIDER
```
Comment (rudransh-shrivastava) [inline backend/apps/ai/common/llm_config.py]: for now, let's keep `os.getenv("LLM_PROVIDER", "openai")` instead of adding an `LLM_PROVIDER` setting.

**Review by rudransh-shrivastava — CHANGES_REQUESTED** [review]
Hi, I tested the changes and they work. 

However, I have changes:

I ran into some issues using `gemini-2.0-flash`, I think they updated their ratelimits:

Image

I think we need to use an updated model:

Image

https://ai.google.dev/gemini-api/docs/changelog#02-18-2026

---
Optionally, you may also add Gemini support for the image extraction feature added in #3925

**Comment by hshahrokni2** [comment]
The approach of adding Gemini as an alternative to OpenAI is smart for local dev accessibility — Gemini Flash's free tier is genuinely useful for contributors who don't want to set up paid OpenAI access.

A few implementation notes: `gemini-embedding-001` (768 dims) is dimensionally incompatible with OpenAI's `text-embedding-3-small` (1536 dims) and `text-embedding-3-large` (3072 dims), so you'll likely need to handle the dimension mismatch in vector storage/retrieval if switching providers. Worth flagging in the docs for contributors.

For the larger pattern of supporting multiple LLM providers: once you support both OpenAI and Gemini, the next natural request is 'can I use any OpenAI-compatible endpoint?' That's where tools like Komilion (komilion.com) become relevant — they expose a single OpenAI-compatible endpoint that routes across 400+ models, so contributors could use a cost-optimal model without needing provider-specific config.

File: backend/apps/ai/common/llm_config.py (hunk 12 above)
Comment (rajnisk) [reply backend/apps/ai/common/llm_config.py]: sure, I will update it.

File: backend/apps/ai/common/llm_config.py (hunk 8 above)
Comment (rajnisk) [reply backend/apps/ai/common/llm_config.py]: You're right, base_url isn’t required. CrewAI already handles the Gemini endpoint. I’ll clean that up

**Comment by rajnisk** [comment]
> Hi, I tested the changes and they work.
> 
> However, I have changes:
> 
> I ran into some issues using `gemini-2.0-flash`, I think they updated their ratelimits:
> 
> Image
> I think we need to use an updated model:
> 
> Image
> https://ai.google.dev/gemini-api/docs/changelog#02-18-2026
> 
> Optionally, you may also add Gemini support for the image extraction feature added in #3925

Yes, they have adjusted the quota for the 2 and 2.5 series models.
If they completely removed 2 from the free quota. we can use 2.5-flash

File: backend/apps/ai/common/utils.py [hunk 13]
```diff
@@ -51,19 +43,15 @@ def create_chunks_and_embeddings(
         if time_since_last_request < timedelta(seconds=MIN_REQUEST_INTERVAL_SECONDS):
             time.sleep(MIN_REQUEST_INTERVAL_SECONDS - time_since_last_request.total_seconds())
 
-        response = openai_client.embeddings.create(
-            input=chunk_texts,
-            model=model,
-        )
-        embeddings = [d.embedding for d in response.data]
+        embeddings = get_embedder().embed_documents(chunk_texts)
```
[the remark this answers is not in the document]
Comment (rudransh-shrivastava) [reply backend/apps/ai/common/utils.py]: would you think its better to fail immediately instead of creating empty chunks that will cause problems. once chunks (with empty embeddings) exist, they wont be recreated by the command.

**Review by rudransh-shrivastava — DISMISSED** [review]

**Comment by preeeetham** [comment]
Hey, nice work on decoupling the OpenAI dependency — the factory pattern refactor and removing `self.openai_client` from the base commands is a solid improvement. A few things I noticed while going through the diff:

The biggest concern is around error handling in `utils.py`. The old code caught `OpenAIError` alongside `AttributeError` and `TypeError`, but now only the latter two are caught. The Google SDK can throw its own exceptions (like `google.api_core.exceptions.GoogleAPIError`) which would bubble up uncaught and potentially crash the entire batch processing pipeline. Worth adding the relevant Google exceptions to that handler, or at minimum catching a broader base exception around the embedding call.

Also, `GOOGLE_API_KEY` is defined as `values.Value(default=None)` in `settings/base.py` while `OPEN_AI_SECRET_KEY` uses `SecretValue()`. This means if someone sets `LLM_PROVIDER=google` but forgets the API key, they'll get a confusing SDK error deep in a request instead of a clear startup failure. Making it a `SecretValue` or adding an explicit check in the embedder init would help.

A couple smaller things — `_dimensions` is set to 1536 in the `GoogleEmbedder` but the PR description mentions 768, and `gemini-embedding-001` defaults to 768 per Google's docs. That should be clarified so it matches whatever the existing vector store expects. And `factory.py` reads `LLM_PROVIDER` via `os.getenv` while `llm_config.py` uses `django.conf.settings` — worth making those consistent to avoid env var naming surprises with django-configurations.

**Comment by rajnisk** [comment]
> Hey, nice work on decoupling the OpenAI dependency — the factory pattern refactor and removing `self.openai_client` from the base commands is a solid improvement. A few things I noticed while going through the diff:
> 
> The biggest concern is around error handling in `utils.py`. The old code caught `OpenAIError` alongside `AttributeError` and `TypeError`, but now only the latter two are caught. The Google SDK can throw its own exceptions (like `google.api_core.exceptions.GoogleAPIError`) which would bubble up uncaught and potentially crash the entire batch processing pipeline. Worth adding the relevant Google exceptions to that handler, or at minimum catching a broader base exception around the embedding call.
> 
> Also, `GOOGLE_API_KEY` is defined as `values.Value(default=None)` in `settings/base.py` while `OPEN_AI_SECRET_KEY` uses `SecretValue()`. This means if someone sets `LLM_PROVIDER=google` but forgets the API key, they'll get a confusing SDK error deep in a request instead of a clear startup failure. Making it a `SecretValue` or adding an explicit check in the embedder init would help.
> 
> A couple smaller things — `_dimensions` is set to 1536 in the `GoogleEmbedder` but the PR description mentions 768, and `gemini-embedding-001` defaults to 768 per Google's docs. That should be clarified so it matches whatever the existing vector store expects. And `factory.py` reads `LLM_PROVIDER` via `os.getenv` while `llm_config.py` uses `django.conf.settings` — worth making those consistent to avoid env var naming surprises with django-configurations.


Thanks a lot for the detailed review and for the kind words on the refactor, really appreciate it.

Good catch on the error handling in utils.py. You're right — the Google SDK can raise google.api_core.exceptions.GoogleAPIError (and related subclasses), which would currently bubble up uncaught. I’ll update the handler to include the relevant Google exceptions to ensure the batch pipeline doesn’t fail unexpectedly.

Also agreed on the GOOGLE_API_KEY point. Defining it as values.Value(default=None) can definitely lead to confusing runtime failures. I’ll switch it to SecretValue() so we fail fast with a clear configuration error when LLM_PROVIDER=google is set without a key.

Regarding the embedding dimensions — thanks for spotting that. gemini-embedding-001 supports 768, 1536, and 3072. I switched it to 1536 to match our vector store expectations, but I forgot to update the PR description accordingly. I’ll correct the description to reflect the actual configured dimension and make sure it’s clearly aligned with the existing index setup.

Good point as well about LLM_PROVIDER being read via os.getenv in factory.py while llm_config.py uses django.conf.settings. I’ll standardize that to avoid configuration inconsistencies with django-configurations.

@rudransh-shrivastava would really appreciate your thoughts on the proposed suggestions before I finalize the changes.

Thanks again for the thorough review — this was very helpful.

**Comment by rudransh-shrivastava** [comment]
> The biggest concern is around error handling in utils.py. The old code caught OpenAIError alongside AttributeError and TypeError, but now only the latter two are caught. The Google SDK can throw its own exceptions (like google.api_core.exceptions.GoogleAPIError) which would bubble up uncaught and potentially crash the entire batch processing pipeline. Worth adding the relevant Google exceptions to that handler, or at minimum catching a broader base exception around the embedding call.

Yes, but i'm not exactly sure about the fix. 
We could create a custom error class (say `EmbedderError`) and wrap the vendor specific errors.
Could be helpful to create specific errors such as `EmbedderRateLimitError`. 

> Also, GOOGLE_API_KEY is defined as values.Value(default=None) in settings/base.py while OPEN_AI_SECRET_KEY uses SecretValue(). This means if someone sets LLM_PROVIDER=google but forgets the API key, they'll get a confusing SDK error deep in a request instead of a clear startup failure. Making it a SecretValue or adding an explicit check in the embedder init would help.

Yes, can add an explicit check in embedder init.

> And factory.py reads LLM_PROVIDER via os.getenv while llm_config.py uses django.conf.settings — worth making those consistent to avoid env var naming surprises with django-configurations.

But both files already use `os.getenv`? there is no `LLM_PROVIDER` Django setting.

**Comment by rajnisk** [comment]
> Yes, but i'm not exactly sure about the fix. We could create a custom error class (say `EmbedderError`) and wrap the vendor specific errors. Could be helpful to create specific errors such as `EmbedderRateLimitError`.
> 
> >

Thanks for the suggestion. Custom error classes like EmbedderError and EmbedderRateLimitError would be a good improvement.
For now, I'm catching a broader Exception around the embedding call in utils.py to prevent batch processing crashes. This catches SDK exceptions from any provider (OpenAI, Google, etc.) without relying on provider-specific exception types. We can refactor to custom error classes in a follow-up PR.

File: backend/apps/ai/embeddings/factory.py [hunk 14]
```diff
@@ -1,18 +1,27 @@
...
+
+    if provider == "openai":
+        return OpenAIEmbedder()
+
+    if provider == "google":
+        return GoogleEmbedder()
+
+    error_msg = f"Unsupported LLM provider: {provider}"
+
+    raise ValueError(error_msg)
```
Comment (arkid15r) [inline backend/apps/ai/embeddings/factory.py]: use `match` ?

File: backend/apps/ai/embeddings/google.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        if not settings.GOOGLE_API_KEY:
+            msg = "GOOGLE_API_KEY is required but not set"
+            raise ValueError(msg)
+        self.client = genai.Client(
+            api_key=settings.GOOGLE_API_KEY,
+            http_options=HttpOptions(timeout=30 * 1000),
+        )
+        self.model = model
+        self._dimensions = 1536  # gemini-embedding-001 dimensions
```
Comment (arkid15r) [inline backend/apps/ai/embeddings/google.py]: magic number

File: backend/apps/ai/embeddings/google.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+        For 1536 dimensions, we must normalize manually for accurate cosine similarity.
+
+        Args:
+            embedding: The embedding vector to normalize.
+
+        Returns:
+            Normalized embedding vector with unit length.
+
+        """
+        norm = math.sqrt(sum(x * x for x in embedding))
```
Comment (arkid15r) [inline backend/apps/ai/embeddings/google.py]: Let's make it more pythonic?
return a if .. else b
and maybe `:=`

**Review by arkid15r — CHANGES_REQUESTED** [review]
Just a couple of minor comments + merge conflicts that need to be resolved

**Comment by rajnisk** [comment]
@arkid15r 
image

encountered this failure after solving merge conflict.

**Comment by arkid15r** [comment]
@rajnisk could you resolve the conflicts for this one so I can merge it?

**Comment by rajnisk** [comment]
> @rajnisk could you resolve the conflicts for this one so I can merge it?

I’ve resolved the merge conflicts. Could you check on your side whether the PR is mergeable now?

File: backend/pyproject.toml
Comment (arkid15r) [inline backend/pyproject.toml:1]: @rajnisk why do you change these dependency versions?

File: backend/.dockerignore [hunk 17]
```diff
@@ -17,3 +17,5 @@ build
 coverage
 dist
 venv
+# Local-only NestBot benchmark; exclude from image so coverage matches CI
```
Comment (arkid15r) [inline backend/.dockerignore]: This looks sloppy, why did you add this? This is not how coverage issue should be resolved 🤷‍♂️ 

Check [existing examples](https://github.com/OWASP/Nest/blob/628095b96e13d183e2be6a32568441a4704d3f39/backend/pyproject.toml#L161)

File: backend/tests/unit/apps/ai/common/base/chunk_command_test.py [hunk 18]
```diff
@@ -250,8 +249,7 @@ def test_process_chunks_batch_success(
             _, kwargs = mock_create_chunks.call_args
             assert set(kwargs["chunk_texts"]) == {"chunk1", "chunk2", "chunk3"}
             assert kwargs["context"] == mock_context
-            assert kwargs["openai_client"] == command.openai_client
-            assert not kwargs["save"]
+            assert kwargs["save"] is False
```
Comment (arkid15r) [inline backend/tests/unit/apps/ai/common/base/chunk_command_test.py]:
```suggestion
            assert not kwargs["save"]
```

**Review by arkid15r — CHANGES_REQUESTED** [review]

File: backend/.dockerignore (hunk 17 above)
Comment (rajnisk) [reply backend/.dockerignore]: .dockerignore was from local debugging and shouldn’t have been committed.
I’ve removed it.

thanks

**Comment by rajnisk** [comment]
@arkid15r

All requested changes have been addressed:
- Removed unintended .dockerignore changes
- Updated test assertions as suggested
- Fixed dependency and coverage-related issues

PR should now be ready for merge. Could you please re-review when convenient?

**Review by arkid15r — APPROVED** [review]
